confluent_kafka/schema_registry/protobuf.py

```diff
@@ -0,0 +1,437 @@
+        while True:
+            i = ProtobufDeserializer._read_byte(buf)
+
+            value |= (i & 0x7f) << shift
+            shift += 7
+            if not (i & 0x80):
+                break
+
+        return value
+
+    @staticmethod
+    def _read_byte(buf):
+        """
+        Returns int representation for a byte.
+
+        Args:
+            buf (BytesIO): buffer to read from
+
+        .. _ord:
+            https://docs.python.org/2/library/functions.html#ord
+        """
+        i = buf.read(1)
+        if i == b'':
+            raise EOFError("Unexpected EOF while reading index")
+        return ord(i)
+
+    @staticmethod
+    def _decode_index(buf, expect):
+        """
+        Extracts message index from Schema Registry Protobuf formatted bytes.
+
+        Args:
+            buf (BytesIO): byte buffer
+
+            expect ([int]): expected message index
+
+        Returns:
+            [int]: Message index. Not currently used.
+
+        Raises:
+            SerializationError: If message indexes do not match
+
+        """
+        expect_size = expect[0]
+        size = ProtobufDeserializer._decode_uvarint(buf)
+        # fail fast, this is guaranteed not to be our expected message type.
+        if expect_size != size:
+            raise SerializationError("Message index mismatch. The consumed"
+                                     " Message does not match this"
+                                     " Deserializer's Message type")
+
+        msg_index = [size]
+        for _ in range(size):
+            msg_index.append(ProtobufDeserializer._decode_uvarint(buf))
+
+        if msg_index != expect:
+            raise SerializationError("Message index mismatch. The consumed"
+                                     " Message does not match this"
+                                     " Deserializer's Message type")
+
+        return msg_index
+
+    def __call__(self, value, ctx):
+        """
+        Deserializes Schema Registry formatted Protobuf to Protobuf Message.
+
+        Args:
+            value (bytes): Confluent Schema Registry formatted Protobuf bytes.
+
+            ctx (SerializationContext): Metadata pertaining to the serialization
+                operation.
+
+        Returns:
+            Message: Protobuf Message instance
+
+        Raises:
+            SerializerError: If response payload and expected Message type
+            differ.
+
+        """
+        if value is None:
+            return None
+
+        # SR wire protocol + msg_index length
+        if len(value) < 6:
+            raise SerializationError("message is too small to decode {}".format(value))
+
+        with _ContextStringIO(value) as payload:
+            magic, schema_id = struct.unpack('>bI', payload.read(5))
+            if magic != _MAGIC_BYTE:
+                raise SerializationError(
+                    "message does not start with magic byte")
+
+            # Protobuf Messages are self-describing; no need to query schema
```

Comment:
read schema id here

Reply:
> magic, schema_id = struct.unpack('>bI', payload.read(5)

Reply:
its more efficient to read/write the schema id with the magic byte since they are both known quantities

Reply:
yep, somehow i was blind to the this. good work, carry on!